pennylane_lightning/lightning_qubit/_measurements.py

```diff
@@ -0,0 +1,173 @@
+# Copyright 2018-2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Class implementation for state vector measurements.
+"""
+
+# pylint: disable=import-error, no-name-in-module, ungrouped-imports
+try:
+    from pennylane_lightning.lightning_qubit_ops import (
+        MeasurementsC64,
+        MeasurementsC128,
+    )
+except ImportError:
+    pass
+
+from typing import Callable, List
+import numpy as np
+
+from pennylane.measurements import StateMeasurement, MeasurementProcess, ExpectationMP
+from pennylane.typing import TensorLike, Result
+from pennylane.tape import QuantumScript
+from pennylane.wires import Wires
+
+from pennylane_lightning.core._serialize import QuantumScriptSerializer
+from ._state_vector import LightningStateVector
+
+
+class LightningMeasurements:
+    """Lightning Measurements class
+
+    Measures the state provided by the LightningStateVector class.
+
+    Args:
+        qubit_state(LightningStateVector)
+    """
+
+    def __init__(self, qubit_state: LightningStateVector) -> None:
+        self._qubit_state = qubit_state
+        self.state = qubit_state.state_vector
+        self.dtype = qubit_state.dtype
+        self.measurement_lightning = self._measurement_dtype()(self.state)
+        
+
+    def _measurement_dtype(self):
+        """Binding to Lightning Managed state vector.
+
+        Args:
+            dtype (complex): Data complex type
+
+        Returns: the state vector class
+        """
+        return MeasurementsC64 if self.dtype == np.complex64 else MeasurementsC128
+
+    def state_diagonalizing_gates(self, measurementprocess: StateMeasurement) -> TensorLike:
+        """Apply a measurement to state when the measurement process has an observable with diagonalizing gates.
+            This method will is bypassing the measurement process to default.qubit implementation.
+
+        Args:
+            measurementprocess (StateMeasurement): measurement to apply to the state
+
+        Returns:
+            TensorLike: the result of the measurement
+        """
+        self._qubit_state.apply_operations(measurementprocess.diagonalizing_gates())
```

Comment:
We may potentially want to forbid measurements to have diagonalizing gates at the point the pipeline.  We don't want to end up in a situation where we have the measurements `[qml.probs(op=qml.X(0)), qml.probs(op=qml.Y(0))]`, and then the second one ends up being wrong.

Reply:
How do you propose doing that?

Reply:
```suggestion
       if measurementprocess.diagonalizing_gates():
           raise NotImplementedError("Measurement process {measuremnetprocess} should be diagonalized before submission to `dev.execute`")
```

I'll write up a transform to add to preprocessing for this case.

Reply:
Sounds good!

Reply:
@vincentmr tackled this in #627 by [applying the adjoint](https://github.com/PennyLaneAI/pennylane-lightning/blob/6ea2ee82070d94560d5095a5c160f8b30e0836b0/pennylane_lightning/lightning_qubit/_measurements.py#L151) of the diagonalizing gates after measuring. That is also a good approach to avoid raising an error.

Reply:
Nice suggestion @mudit2812 and @vincentmr. I just pushed a change that implements that. @mudit2812, could you please double-check if it makes sense and mark this conversation as resolved if sensible?